### PR TITLE
Finish MetaQuestionRepository

### PR DESCRIPTION
**Pull request by RoiTief**

### CODE REVIEW

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 
 class MetaQuestionRepository {
     #MetaQuestion
     #Answer
     #Keyword
+    #Appendix
 
     constructor(sequelize) {
         this.#MetaQuestion = defineMetaQuestionModel(sequelize);
         this.#Answer = defineAnswerModel(sequelize);
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:16]: we changed the name to key (keys)

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:1]: I love the way they use to define table relations, really readable

File: src/main/DAL/MetaQuestion/Appendix.js [hunk 2]
```diff
@@ -7,15 +7,16 @@ function defineAppendixModel(sequelize) {
     return sequelize.modelManager.models.some(model => model.name === 'Appendix') ?
         sequelize.models.Appendix :
         sequelize.define('Appendix', {
-                title: {
+                tag: {
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/Appendix.js:10]: Is tag should represent id? and if so why not to call it id?

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 3]
```diff
@@ -21,7 +26,38 @@ class MetaQuestionRepository {
         this.#MetaQuestion.belongsToMany(this.#Keyword, {through: 'MetaQuestionKeyword', as: 'keywords'});
         this.#Keyword.belongsToMany(this.#MetaQuestion, {through: 'MetaQuestionKeyword', as: 'metaQuestions'});
 
-        sequelize.sync({alter: true}); // cleans the DB
+        // MQ - Appendix
+        this.#Appendix.hasMany(this.#MetaQuestion, {as: 'metaQuestions', foreignKey: 'appendixTag'});
+        this.#MetaQuestion.belongsTo(this.#Appendix, {as: 'appendix', foreignKey: { name: 'appendixTag', allowNull: true }});
+
+        // Appendix - Keyword
+        this.#Appendix.belongsToMany(this.#Keyword, {through: 'AppendixKeyword', as: 'keywords'});
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:34]: cant we create appendix and metaquestion without kewords?

cant we create a keyword that will assign only to metaquestion without appendix?

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 4]
```diff
@@ -44,12 +80,78 @@ class MetaQuestionRepository {
...
+     * @param keywords Array of keywords: string[]
+     * @returns The updated question.
+     */
+    async addKeywordsToQuestion(qId, keywords) {
+        const metaQuestion = await this.getMetaQuestion(qId);
+        const dbKeywords = await this.#getDbKeywords(keywords)
+        await metaQuestion.addKeywords(dbKeywords);
+        await metaQuestion.reload();
+        return metaQuestion;
+    }
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:103]: what if keywords is not already saved in the db ? does getDbKeywords save them?

metaQuestion.addKeywords add them ?

Same goes with addKeywordsToAppendix....

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 5]
```diff
@@ -44,12 +80,78 @@ class MetaQuestionRepository {
      * @returns The updated question.
      */
     async addAnswersToQuestion(qId, answers) {
-        await answers.forEach(answer => {
+        answers.forEach(answer => {
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:83]: I think you are the one who told me to change from answers to keys

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 6]
```diff
@@ -44,12 +80,78 @@ class MetaQuestionRepository {
...
+        return appendix;
+    }
+
+    /**
+     * @param tag Tag of the requested Appendix.
+     * @return The requested Appendix.
+     */
+    async getAppendix(tag) {
+        return await this.#Appendix.findByPk(tag, {
+            include: [
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:125]: amazing syntax

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js [hunk 7]
```diff
@@ -84,6 +181,10 @@ class MetaQuestionRepository {
                     model: this.#Keyword,
                     as: 'keywords',
                 },
+                {
+                    model: this.#Appendix,
+                    as: 'appendix',
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestionRepository.js:186]: required: false,

File: src/main/DAL/MetaQuestion/MetaQuestion.js [hunk 8]
```diff
@@ -5,25 +5,14 @@ const {META_QUESTION_TYPES} = require("../../Enums");
...
-            type: DataTypes.ENUM,
-            values: Object.values(META_QUESTION_TYPES),
-            allowNull: false
-        }
-    }, {
-       sequelize,
-       modelName: 'MetaQuestion',
-    });
-    return sequelize.models.MetaQuestion;
+    return sequelize.modelManager.models.some(model => model.name === 'MetaQuestion') ?
```
Comment (MayElnathan) [inline src/main/DAL/MetaQuestion/MetaQuestion.js:8]: we said that metaquestion should have id

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 1 above)
Comment (MayElnathan) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:16]: Isent there #Distructors?

File: src/main/DAL/MetaQuestion/Appendix.js (hunk 2 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/Appendix.js:10]: sole reason is that's how Mayer originally sectioned the appendix: tag, title, content
where tag is unique

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 1 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:16]: what are you talking about? `key` is a type of an answer, not a type of appendix

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 3 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:34]: 1. yes we can, pass the empty array [] and it'll be created without keywords
2. yes we can, yes we can. this relation is defined this way solely for the option of recycling keywords

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 4 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:103]: yes exactly that. getDbKeywords creates rows for the given keywords (unless they already exist). 
once you have an array of DB elements the method `addKeywords` (which is generated through the relation definition automatically) creates the necessary connections in the DB

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 5 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:83]: `correctAnswers` changed to `keys`
$Answers = Keys \cup Distractors$
Why do we work with the superset `Answers` then?
An answer might be wrongly tagged as a `key` while in actuality be wrong in relation to the `MetaQuestion` it is connected to: 
- If we work with the `Answer` model we simply switch the tag from `key` to `distractor. 
- If we work with the `Key` and `Distractor` model, in order to switch the tagging we will need to remove the instance from the `Keys` table and create the instance as a row in the `Distractors` table.

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 6 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:125]: This is sequelize, not me....
They did do an amazing job

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 7 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:186]: if the MQ is not connected to an Appendix this field will be `null`. as defined by the relationship.

File: src/main/DAL/MetaQuestion/MetaQuestion.js (hunk 8 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestion.js:8]: they do have an ID
By not defining any Primary-Key, sequelize gives an enumerated id column to the table.

Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestion.js:8]: check out the tests

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 1 above)
Comment (MayElnathan) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:16]: I talked about #Asnwers

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 7 above)
Comment (MayElnathan) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:186]: we dont want to work with null.

you so what it did to us...

make sure it is undefined

Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:186]: This is too much logic for DAL. 
I would actually like to get a good amount of information (even maybe a little extra) from DAL layer and then Business layer can manipulate it in whatever form we want. that way DAL layer won't need many refactors.

File: src/main/DAL/MetaQuestion/MetaQuestionRepository.js (hunk 1 above)
Comment (RoiTief) [reply src/main/DAL/MetaQuestion/MetaQuestionRepository.js:16]: answered in different conversation

**Review by MayElnathan — APPROVED** [review]

**Review by MayElnathan — APPROVED** [review]

**Pull request merged**